Core/Consumers/AbstractAsyncConsumer.php

```diff
@@ -0,0 +1,214 @@
+    protected $sleep = true;
+
+    /**
+     * Holds the amount of time it took to process a message once the callback is triggered. Due to the asynchronicity
+     * nature of this consumer, it's not possible to measure how long it takes to get a message from the queue and send
+     * it to the callback
+     *
+     * @var int
+     */
+    protected $consumptionDuration = 0;
+
+    /**
+     * Initializes the consumer for a given endpoint.
+     *
+     * @param EndpointInterface $endpoint
+     */
+    abstract protected function initialize(EndpointInterface $endpoint);
+
+    /**
+     * Cleans up and closes connections before shutting down.
+     *
+     * @param EndpointInterface $endpoint
+     *
+     * @return mixed
+     */
+    abstract protected function cleanUp(EndpointInterface $endpoint);
+
+    /**
+     * After the execution of this method, it will be considered that the message was successfully handled,
+     * therefore, if there was any problem, an exception must be thrown and not continue. This is important to ensure
+     * the Message Delivery Guarantee.
+     *
+     * @param EndpointInterface $endpoint
+     * @param MessageInterface $message
+     */
+    protected function process(EndpointInterface $endpoint, MessageInterface $message)
+    {
+        $endpoint->handle($message);
+    }
+
+    /**
+     * This function is called to confirm that a message was successfully handled. Until this point, the message should
+     * not be removed from the source Endpoint, this is very important to ensure the Message delivery guarantee.
+     *
+     * @param EndpointInterface $endpoint
+     * @param $message
+     *
+     * @return MessageInterface
+     */
+    abstract protected function confirmMessage(EndpointInterface $endpoint, MessageInterface $message);
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \Exception
+     */
+    public function consume(EndpointInterface $endpoint)
+    {
+        $this->initialize($endpoint);
+
+        $callback = $this->callback($endpoint);
+        $this->asyncConsume($endpoint, $callback);
+
+        while (!$this->shouldStop() && $this->sleep = true) {
+            try {
+                $this->waitNoBlock($endpoint);
+
+                if ($this->sleep) {
+                    usleep(self::SLEEP_DURATION);
+                }
+            } catch (\Exception $exception) {
+                if (!$this->stop) {
+                    throw $exception;
+                }
+            }
+        }
+
+        $this->cleanUp($endpoint);
+    }
+
+    /**
+     * Callback function to be triggered when a message is received. Is in charge of triggering the processing,
+     * confirming the message if no exception was thrown, reducing the expiration count and preventing the consumer
+     * from sleeping in the next round.
+     *
+     * @param EndpointInterface $endpoint
+     *
+     * @return \Closure
+     */
+    public function callback(EndpointInterface $endpoint): callable
```

Comment:
Is it needed this method to be public?

Reply:
Function protected. Tests we adjusted accordingly.

Reply:
I'll thumbs up myself just for the sake of it :D 

:+1: